design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,188 @@
+Deployments which are already updated. OSD Deployments which need updated will have a
+`rook-version` label that does not match the current version of the Rook operator and a
+`ceph-version` label that does not match the current Ceph version being deployed.
+
+Go channels are likely not the best option for the update queue since it may be necessary to select
+individual items from the queue at any position and not merely grab the top/bottom item from the
+queue.
+
+The minimal information each item in the queue needs is only the OSD ID. The OSD Deployment managed
+by Rook can easily be inferred from the OSD ID.
+
+### User configuration
+
+#### Modifications to `CephCluster` CRD
+Establish a new `updatePolicy` section in the `CephCluster` `spec`. In this section, users can
+set options for how OSDs should be updated in parallel. Additionally, we can move some existing
+one-off configs related to updates to this section for better coherence. This also allows for
+a natural location where future update options can be added.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+# ...
+spec:
+  # ...
+  # Move these to the new updatePolicy but keep them here for backwards compatibility.
+  # These can be marked deprecated, but do not remove them until CephCluster CRD v2.
+  skipUpgradeChecks:
+  continueUpgradeAfterChecksEvenIfNotHealthy:
+  removeOSDsIfOutAndSafeToDestroy:
+
+  # Specify policies related to updating the Ceph cluster and its components. This applies to
+  # minor updates as well as upgrades.
+  updatePolicy:
+    # skipUpgradeChecks is merely relocated from spec
+    skipUpgradeChecks: <bool, default=false>
+
+    # continueUpgradeAfterChecksEvenIfNotHealthy is merely relocated from spec
+    continueUpgradeAfterChecksEvenIfNotHealthy: <bool, default=false, relocated from spec>
+
+    # allow for future additions to updatePolicy like healthErrorsToIgnore
+
+    # Update policy for OSDs.
+    osds:
+      # removeIfOutAndSafeToDestroy is merely relocated from spec (removeOSDsIfOutAndSafeToRemove)
+      removeIfOutAndSafeToDestroy: <bool, default=false>
+
+      # Max number of OSDs in the cluster to update at once. Rook will try to update this many OSDs 
+      # at once if it is safe to do so. It will update fewer OSDs at once if it would be unsafe to 
+      # update maxInParallelPerCluster at once. This can be a discrete number or a percentage of 
+      # total OSDs in the Ceph cluster.
+      # Rook defaults to updating 15% of OSDs in the cluster simultaneously if this value is unset.
+      # Inspired by Kubernetes apps/v1 RollingUpdateDeployment.MaxUnavailable.
+      # Note: I think we can hide the information about CRUSH from the user since it is not
+      #       necessary for them to understand that complexity.
+      maxInParallelPerCluster: <k8s.io/apimachinery/pkg/util/intstr.intOrString, default=15%>
+```
+
+Default `maxInParallelPerCluster`: Ceph defaults to keeping 3 replicas of an item or 2+1 erasure
```

Comment:
Per discussion today, the proposal is to use the data type in the cluster CR to allow either a number or percentage, but only start by allowing a number, right?

Reply:
I don't see a lot of reason to not implement both from the start. Calculating an int value to pass to `ceph osd ok-to-stop --max=<number>` isn't difficult since we just have to count the number of OSD Deployments for the cluster and multiply by the percentage.

For the example manifests, I'm happy to give examples with just numbers since I agree that is easier for users to comprehend at first. We can even set the default value (internal to Rook) to be 10 or 20 like Sage mentioned.